libass/ass_bitmap.c

```diff
@@ -215,9 +219,23 @@ void ass_fix_outline(Bitmap *bm_g, Bitmap *bm_o)
     uint8_t *g = bm_g->buffer + (t - bm_g->top) * bm_g->stride + (l - bm_g->left);
     uint8_t *o = bm_o->buffer + (t - bm_o->top) * bm_o->stride + (l - bm_o->left);
 
+    // Use a number just above 65025 to ensure den > 0 and we don't divide
+    // by zero. This allows us to save a branch and assist autovectorization.
+    // The cost (which we accept) with IEEE float32 is that just two out
+    // of the 33 million possible operand tuples round in the "wrong"
+    // direction (to the more distant nearby integer) after evaluating
+    // to almost exactly an integer and a half. (Three more round up
+    // to 1 from 0.5; all other tuples round to nearest, half to zero.)
+    const float over65025 = nextafterf(65025, 1e6f);
+
     for (int32_t y = 0; y < b - t; y++) {
-        for (int32_t x = 0; x < r - l; x++)
-            o[x] = (o[x] > g[x]) ? o[x] - (g[x] / 2) : 0;
+        for (int32_t x = 0; x < r - l; x++) {
+            int num = blurred ?
+                o[x] * (255 - g[x]) * 255 :
+                FFMAX(o[x] - g[x], 0) * 65025;
+            float den = over65025 - alpha_g * g[x];
+            o[x] = num / den + 0.5f;
```

Comment:
You can match the integer version perfectly, at negligible perf impact, like this:
```suggestion
            float den = 65025 - alpha_g * g[x];
            o[x] = FFMIN(num / den, 255.f) + 0.5f;
```

Reply:
This assumes we have IEEE… uh… no, does this even work?

If `den` is 0, then `num` is 0, too. With IEEE float, `num / den` is NaN, and this `FFMIN` returns NaN (if I’m reading its definition correctly—as much as it doesn’t expect NaNs as inputs in the first place).

Reply:
> and not-zero in practice

Actually, it probably _is_ zero, at least on x86? The conversion probably goes through int32, giving `INT32_MIN` on x86, and then truncated to int8, giving zero?

In that case, this is actually a valid trick to use in hand-written assembler code that could target any C implementation that explicitly guards against division by zero.

But in C itself, no, I don’t think this is allowed. In fact, §6.5.5 <i>Multiplicative operators</i> explicitly declares any division by zero UB in any case.

Reply:
Actually, instead of FFMIN we could use
```c
float den = 65025 - alpha_g * g[x];
o[x] = num / (den + 1e-30f) + 0.5f;
```
This should have a significantly lower impact than over65025.

Reply:
Or we could do `float divres = (num / den); o[x] = (divres >= 0.f ? divres : 0.f) + 0.5f;`.

Are we aware of any real targets we care about that don't provide IEEE floats? I'm perfectly comfortable having those as a requirement.

Reply:
I don’t know, but I’m not a specialist in other platforms. My only worthless contribution is that IIRC GPUs don’t (or didn’t), but I’d be surprised if libass could be compiled for a GPU.

But for starters, `-ffast-math`, which e. g. MrSmile uses for his own builds, is a common way to throw NaN explicitly out the window. It _probably_ wouldn’t affect this line simply because the most trivial thing for the compiler to do is exactly the thing we find most desirable, but I’m not a fan of “probably usually maybe correct but explicitly unguaranteed” code. This function doesn’t seem worth bumping libass’s requirements to include IEEE compliance, and it might well harm us in other places (I assume MrSmile uses `-ffast-math` not for nothing).

For another thing, neither GCC nor Clang (nor MSVC) actually fully support Annex F nor define `__STDC_IEC_559__` (although some platforms have system headers that define it instead—which isn’t correct), so this remains technically _undefined_ behaviour. Indeed, Clang even used to complain about it with `-fsanitize=undefined`. But again, it’s true that, in practice, their optimizers do ensure 0/0=NaN, except when this is disabled via `-ffast-math` etc. (and so Clang’s UBSan excludes this check by default since Clang 9). [[17000](https://bugs.llvm.org/show_bug.cgi?id=17000)] [[19535](https://bugs.llvm.org/show_bug.cgi?id=19535)]

> `den + 1e-30f`

I like this idea (if we stick to dividing every pixel).

Reply:
> This should have a significantly lower impact than over65025.

Remind me: what is the impact of over65025 in this context? The `nextafterf` call itself? But it happens only once, whereas `+ 1e-30f` would have to happen in each iteration.

Reply:
iirc the call to `nextafterf` was _very_ expensive, while the constant adds are quite cheap? But also, the `nextafterf` doesn't quite match the int implementation in all cases, while the add does.

Reply:
> Remind me: what is the impact of over65025 in this context?

Result change. Basically `+ 1e-30` only matters for `den == 0`, while `over65025` can shift values even in nonzero case.